src: implement whatwg's URLPattern spec

anonrig opened the pull request:
Co-authored-by: Daniel Lemire (@lemire)

## Notes:

- Ada now requires C++20
- URLPattern is exposed through node:url module

## TODOs

- [x] Removed exception flag requirement
- [x] Pass all web-platform tests
- [x] Release Ada v3 before landing this PR
- [x] Make sure to split all changes to multiple commits
- [x] Add @lemire as co-author to all commits
- [x] Land upstream pull-request https://github.com/ada-url/ada/pull/785
- [x] Add documentation for global and node:url module declarations.

cc @nodejs/cpp-reviewers 

Fixes https://github.com/nodejs/node/issues/40844

anonrig commented on test/parallel/test-shadow-realm-globals.js:
```diff
@@ -24,4 +24,4 @@ for (const item of items) {
 }
 
-assert.deepStrictEqual(leaks, []);
+assert.deepStrictEqual(leaks, ['URLPattern']);
```
@legendecas Is this correct? How can I make sure it is not exposed?

legendecas replied:
The interfaces exposed in `lib/internal/bootstrap/web/exposed-wildcard.js` should be annotated in the WebIDL as `[Exposed=*]`.

https://urlpattern.spec.whatwg.org/#urlpattern-class defines that `URLPattern` is annotated as `[Exposed=(Window,Worker)]`, so it should be exposed in `lib/internal/bootstrap/web/exposed-window-or-worker.js`.

legendecas replied:
This reminds why not exposing it as `[Exposed=*]`. https://github.com/whatwg/urlpattern/pull/236 is the pending work to do it.

anonrig replied:
Thank you. Extremely helpful as always!

targos commented:
> Ada now enables exceptions just like UV and V8

Can you elaborate? libuv is a C library so I don't think exceptions exist there, and I'm pretty sure V8 is built with exceptions disabled.

anonrig commented:
> > Ada now enables exceptions just like UV and V8
> 
> Can you elaborate? libuv is a C library so I don't think exceptions exist there, and I'm pretty sure V8 is built with exceptions disabled.

My bad UV does not enable exceptions. Referencing v8.gyp file:

```
{
  'target_name': 'torque_base',
  'type': 'static_library',
  'toolsets': ['host', 'target'],
  'sources': [
    '<!@pymod_do_main(GN-scraper "<(V8_ROOT)/BUILD.gn"  "\\"torque_base.*?sources = ")',
  ],
  'dependencies': [
    'v8_shared_internal_headers',
    'v8_libbase',
  ],
  'defines!': [
    '_HAS_EXCEPTIONS=0',
    'BUILDING_V8_SHARED=1',
  ],
  'cflags_cc!': ['-fno-exceptions'],
  'cflags_cc': ['-fexceptions'],
  'xcode_settings': {
    'GCC_ENABLE_CPP_EXCEPTIONS': 'YES',  # -fexceptions
  },
  'msvs_settings': {
    'VCCLCompilerTool': {
      'RuntimeTypeInfo': 'true',
      'ExceptionHandling': 1,
    },
  },
}
```

targos commented:
This is not really V8. It's a build-time executable (torque) used to generate code for V8

jasnell commented on src/node_url_pattern.cc:
```diff
@@ -0,0 +1,571 @@
+URLPattern::URLPattern(Environment* env,
+                       Local<Object> object,
+                       ada::url_pattern&& url_pattern)
+    : BaseObject(env, object) {
+  MakeWeak();
+  url_pattern_ = std::move(url_pattern);
```
```suggestion
    : BaseObject(env, object),
      url_pattern_(std::move(url_pattern)) {
  MakeWeak();
```

jasnell commented on src/node_url_pattern.cc:
```diff
@@ -0,0 +1,571 @@
+      ToV8Value(env->context(), result.input).ToLocalChecked(),
+      ToV8Value(env->context(), result.groups).ToLocalChecked(),
+  };
+  return Object::New(isolate, Null(isolate), names, values, 2);
```
```suggestion
  return Object::New(isolate, Null(isolate), names, values, arraysize(names));
```

here and the other places you`re using that pattern

jasnell commented on src/node_url_pattern.cc:
```diff
@@ -0,0 +1,571 @@
+  ada::url_pattern_options options{};
+  Local<Value> ignore_case;
+  if (obj->Get(env->context(),
+               FIXED_ONE_BYTE_STRING(env->isolate(), "ignoreCase"))
```
consider adding to env-properties.h

jasnell commented on src/node_url_pattern.cc:
```diff
@@ -0,0 +1,571 @@
+
+MaybeLocal<Value> URLPattern::Hash() const {
+  auto context = env()->context();
+  return ToV8Value(context, url_pattern_.get_hash());
```
I think the key challenge here is that this will copy the string on every call. Any chance of memoizing the string once created.

anonrig replied:
I'll follow up on a different PR for this

jasnell commented on src/node_url_pattern.cc:
```diff
@@ -0,0 +1,571 @@
+URLPattern::URLPattern(Environment* env,
+                       Local<Object> object,
+                       ada::url_pattern&& url_pattern)
+    : BaseObject(env, object) {
```
We likely should introduce this as experimental in the first release, even if it graduates from experimental quickly. There should likely be a warning emitted on the first construction.

anonrig replied:
AFAIK: There is no easy way to emit an experimental warning in C++ that can be dismissed using the CLI command. For now, I have made it experimental on the nodejs doc.

jasnell commented on lib/internal/url.js:
```diff
@@ -1571,6 +1572,7 @@ module.exports = {
   toPathIfFileURL,
   installObjectURLMethods,
   URL,
+  URLPattern,
```
Needs docs added.

jasnell commented on src/node_url_pattern.cc:
```diff
@@ -0,0 +1,571 @@
+  URLPattern* url_pattern;
+  ASSIGN_OR_RETURN_UNWRAP(&url_pattern, info.This());
+  Local<Value> result;
+  if (!url_pattern->Protocol().ToLocal(&result)) {
+    return;
+  }
+  info.GetReturnValue().Set(result);
```
```suggestion
  if (url_pattern->Protocol().ToLocal(&result)) {
    info.GetReturnValue().Set(result);
  }
```

a bit more compact to invert the checks on these.

jasnell commented on src/node_url_pattern.cc:
```diff
@@ -0,0 +1,571 @@
+    return Null(env->isolate());
+  }
+  env->ThrowTypeError("Failed to exec URLPattern");
+  return Undefined(env->isolate());
```
when scheduling a throw, the idiomatic return should be empty.

```suggestion
  return MaybeLocal<Value>();
```

jasnell commented on src/node_url_pattern.cc:
```diff
@@ -0,0 +1,571 @@
+
+  if (!url_pattern) {
+    env->ThrowTypeError("Failed to construct URLPattern");
+  }
```
shouldn't this return? I don't think this exits early as is.

jasnell commented on src/node_url_pattern.cc:
```diff
@@ -0,0 +1,571 @@
+void URLPattern::New(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  CHECK(args.IsConstructCall());
```
since this constructor is exposed directly to users, this should throw an exception rather than abort

jasnell replied:
```suggestion
  if (!args.IsConstructCall()) {
    return THROW_........
  }
```

Replace the CHECK here with an appropriate error.

jasnell commented:
Can you also include a fairly simple benchmark?

lukeed commented on src/node_url_pattern.cc:
```diff
@@ -0,0 +1,571 @@
+    BufferValue input_buffer(env->isolate(), args[0]);
+    CHECK_NOT_NULL(*input_buffer);
+    input = input_buffer.ToString();
+  } else if (args[0]->IsObject()) {
```
It’s not part of spec but can we add a “is URL instance” fast path here? To avoid re-normalizing each URL property (as with the JsObject route)

anonrig replied:
Yes, we can. I haven't started working on optimizations yet due to the inconsistencies between the spec and WPT.

mcollina reviewed (changes requested):
I don’t think this is a good pattern to land in Node.js. Specifically, a server using this will create one per route and iterate in a loop. This will be _slow_, specifically if you need to match the last of the list.

(This feedback was provided when URLPattern was standardized and essentially ignored).

For this to be _useful_, we would need to have a Node.js-specific API to organize these URLPattern in a radix prefix trie and actually do the matching all at once.

I can possibly be persuaded that we need this for Web platform compatibility, but it’s not that popular either (unlike fetch()).

mcollina commented:
@jasnell I’ll try to build this and get a benchmark going against the ecosystem routers.

anonrig commented:
> @jasnell I’ll try to build this and get a benchmark going against the ecosystem routers.

Right now, this pull-request does not pass WPT, and not at all optimized. Any benchmarks will not be beneficial.

jasnell commented:
> I can possibly be persuaded that we need this for Web platform compatibility, but it’s not that popular either (unlike fetch()).

It's not a great spec but it is implemented in Deno, Bun, Workers, and browsers. We're looking at adding it to the WinterTC Common Minimum set. Node.js' lack of support is really the only thing holding that up right now. While I recognize that you don't like this particular spec, I think compat and consistency with other runtimes is a worthwhile goal here. We should land this then keep following up with optimizations and improvements.

jasnell commented:
@nodejs/tsc ... please review and weigh in. There is definite demand for this API as we'e had a number of folks ask over the years. It's also implemented in other runtimes and is a web platform API standard. On the downside it's not the best pattern or most performant approach. Ideally we'd be able to resolve without referring to a vote.

marco-ippolito commented:
I'm +1 on adding it to Node.js, I like more WP compatibility and I feel like it's an api that could be useful for libraries.
I will not comment on the implementation details since I'm not an expert on the subject

ronag reviewed (commented):
We have landed plenty of inefficient Web API's. Don't see why this would be different. I'm of the consistent opinion that we should implement them but strongly discourage their use for any performance sensitive code.

mon-jai commented:
Just out of curiosity, why can't we port the Chromium's implementation but instead need to implement this ourselves? (I am not a maintainer)

jasnell commented:
> Just out of curiosity, why can't we port the Chromium's implementation but instead need to reimplement this ...

The chromium implementation depends on quite a few chromium/blink internals that we don't have here. That kind of reuse is difficult to just make happen. This ada-based implementation is standalone, has no other dependencies, etc. It's actually easier overall just to write it from scratch than to repurpose that other impl.

atlowChemi commented on typings/internalBinding/url_pattern.d.ts:
```diff
@@ -0,0 +1,20 @@
+export class URLPattern {
+    protocol: string
```
indentation is off
```suggestion
  protocol: string
```

domenic commented:
One worry worth highlighting here is that, IIUC, the architecture of this PR will use a separate regexp library for the regexps that show up in the `URLPattern` constructor, vs. the ones that show up in the `RegExp` constructor.

That could be confusing for developers, as I assume the feature overlap will not be the same. For example, I wonder what results you would get after this PR for:

```js
new URLPattern("https://example.com/:id([\\p{Decimal_Number}--[1-9]])").test("https://example.com/0");

new RegExp("[\\p{Decimal_Number}--[1-9]]", "v").test("0");
```

In Chromium we have architected our `URLPattern` implementation into two parts: one, [liburlpattern](https://source.chromium.org/chromium/chromium/src/+/main:third_party/liburlpattern/;l=1?q=liburlpattern&sq=&ss=chromium), which is meant to have very few dependencies, and largely works to produce regexp pattern strings. The second part, in [blink/renderer/core/url_pattern](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/url_pattern/), implements more parts of the spec, including [bridging to V8's regexp engine](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/url_pattern/url_pattern_component.cc;l=223-225;drc=01705af75073f34800743f8e2cecf8b317d02881).

I suspect a similar architecture would not fit that well for you, given the different boundaries between Ada / Node.js. But maybe Ada could have some sort of regexp-matcher-callback system which would allow the use of V8's regexps instead of `std::regex`.

anonrig commented:
> I suspect a similar architecture would not fit that well for you, given the different boundaries between Ada / Node.js. But maybe Ada could have some sort of regexp-matcher-callback system which would allow the use of V8's regexps instead of std::regex.

You're right. I think we should also wait for the web-platform tests to be fixed or discussed before releasing an URLPattern implementation that is spec compliant but not Chromium compliant.

jasnell commented on src/node_url_pattern.h:
```diff
@@ -0,0 +1,104 @@
+
+namespace node::url_pattern {
+
+class URLPatternRegexProvider {
```
Nit: Some code comments explaining what this is would be helpful for others coming into this code later

jasnell commented on src/node_url_pattern.cc:
```diff
@@ -0,0 +1,714 @@
+    return;
+  }
+  info.GetReturnValue().Set(result);
+}
```
the way these are defined, the property is going to be reserialized as a new string on every call. Should probably memoize these to avoid the extraneous/duplicative copies.

anonrig replied:
Any suggestions on how to memoize with least amount of code and highest amount of memory safety possible?

jasnell replied:
You would need to have `v8::Global<v8::String>` member fields to cache the values. Or, maybe use `SetLazyDataProperty()` ... not sure if the latter would make it spec compliant, tho, as I believe it makes the properties instance properties rather than prototype properties.

anonrig commented on src/node_url_pattern.cc:
```diff
@@ -0,0 +1,714 @@
+}
+
+void URLPattern::MemoryInfo(MemoryTracker* tracker) const {
+  // TODO(anonrig): Implement this.
```
I'm not sure how to properly set the memory of a url_pattern. Any suggestions?

jasnell replied:
Just estimate, it does not need to be exact. This information is used when generating a heap snapshot so it's largely informational.

anonrig commented on src/node_url_pattern.cc:
```diff
@@ -0,0 +1,714 @@
+  Local<Value> ignore_case;
+  if (obj->Get(env->context(),
+               FIXED_ONE_BYTE_STRING(env->isolate(), "ignoreCase"))
+          .ToLocal(&ignore_case)) {
```
Oops. Nice catch.

anonrig replied:
if it can't get it, we should fail since it is not required. I think this is unnecessary.

anonrig commented:
@nodejs/tsc @nodejs/build @nodejs/platform-macos I can not land this pull-request due to the old macOS infrastructure. This is currently the only blocker for this pull-request.

```
19:50:46 ../deps/ada/ada.h:8457:10: fatal error: 'ranges' file not found
19:50:46 #include <ranges>
19:50:46          ^~~~~~~~
19:50:46   /usr/local/bin/ccache cc -o /Users/iojs/build/workspace/node-test-commit-osx/nodes/osx11-
```

mcollina commented:
What's the status with WPTs?
Did you perform any optimizations?

jasnell commented on src/node_url_pattern.cc:
```diff
@@ -0,0 +1,705 @@
+using v8::Local;
+using v8::LocalVector;
+using v8::MaybeLocal;
+using v8::Name;
```
```suggestion
using v8::Name;
using v8::NewStringType;
```

jasnell commented on src/node_url_pattern.cc:
```diff
@@ -0,0 +1,705 @@
+
+  Local<String> local_pattern;
+  if (!String::NewFromUtf8(
+           isolate, pattern.data(), v8::NewStringType::kNormal, pattern.size())
```
```suggestion
           isolate, pattern.data(), NewStringType::kNormal, pattern.size())
```

jasnell commented on src/node_url_pattern.cc:
```diff
@@ -0,0 +1,705 @@
+  auto env = Environment::GetCurrent(isolate);
+  Local<String> local_input;
+  if (!String::NewFromUtf8(
+           isolate, input.data(), v8::NewStringType::kNormal, input.size())
```
```suggestion
           isolate, input.data(), NewStringType::kNormal, input.size())
```

jasnell commented on src/node_url_pattern.cc:
```diff
@@ -0,0 +1,705 @@
+
+bool URLPatternRegexProvider::regex_match(std::string_view input,
+                                          const regex_type& pattern) {
+  auto isolate = Isolate::GetCurrent();
```
To be extra safe here, you might add a `DCHECK_NOT_NULL(isolate)` for these.

legendecas replied:
`Isolate::GetCurrent()` already asserts the isolate is entered and this function does not return a nullptr. There is an [`Isolate::TryGetCurrent()`](https://source.chromium.org/chromium/chromium/src/+/main:v8/src/api/api.cc;l=10064-10072;drc=27d34700b83f381c62e3a348de2e6dfdc08364b8;bpv=1;bpt=1) which does not abort if an isolate was not entered but I guess this is not the case.

jasnell commented on src/node_url_pattern.cc:
```diff
@@ -0,0 +1,705 @@
+      CHECK(!options.has_value());
+      options = URLPatternOptions::FromJsObject(env, args[0].As<Object>());
+    } else {
+      env->ThrowTypeError("baseURL or options must be provided");
```
Use `ERR_MISSING_ARGS`?

jasnell commented on src/node_url_pattern.cc:
```diff
@@ -0,0 +1,705 @@
+  } else if (args[0]->IsObject()) {
+    init = URLPatternInit::FromJsObject(env, args[0].As<Object>());
+  } else {
+    env->ThrowTypeError("input must be an object or a string.");
```
Use `ERR_INVALID_ARG_TYPE`?

jasnell commented on src/node_url_pattern.cc:
```diff
@@ -0,0 +1,705 @@
+    }
+    return Null(env->isolate());
+  }
+  env->ThrowTypeError("Failed to exec URLPattern");
```
These throws (here and the "Failed to test ..." one below) need proper error codes.

anonrig replied:
Any suggestion on error code naming?

anonrig commented:
> What's the status with WPTs? 

There is currently 19 failing tests, and 6 of them are invalid, and needs to be fixed in WPT. I have an open PR: https://github.com/web-platform-tests/wpt/pull/49782

> Did you perform any optimizations?

Not at the moment. My goal is to pass almost all WPT and land this pull-request before optimizing it. Passing all WPT will give us the confidence to optimize more aggressively.

anonrig commented:
cc @nodejs/cpp-reviewers @nodejs/url can you review? we now pass all valid WPT.

jasnell commented on doc/api/globals.md:
```diff
@@ -1145,6 +1145,16 @@ added: v10.0.0
+
+<!-- YAML
+added: REPLACEME
+-->
```
Only remaining nit... this should be added as Experimental initially.

anonrig commented:
@nodejs/platform-windows @lemire can you help me identify and fix the build error on windows?

lemire commented:
@anonrig On my todo. I will have a look.

anonrig commented:
@targos I've removed the exception requirement of this pull-request. We no longer use std::regex and I've put it under a flag where it's not enabled for Node.js builds: https://github.com/ada-url/ada/pull/853

StefanStojanovic commented:
> @nodejs/platform-windows @lemire can you help me identify and fix the build error on windows?

From what I see it's all ada related. Since we have collaborators that are actively working on that project I'll leave it to them, but feel free to reach out if any assistance is needed.

addaleax commented on src/node_url_pattern.cc:
```diff
@@ -0,0 +1,791 @@
+    input = ada::url_pattern_init{};
+  } else if (args[0]->IsString()) {
+    Utf8Value input_value(env->isolate(), args[0].As<String>());
+    input_base = input_value.ToString();
```
small suggestion: We could just move the `Utf8Value` into the outer scope as a `std::optional<Utf8Value>` and skip the intermediate `std::string` then?

Similar recommendation elsewhere :)

anonrig replied:
Nice catch. Thank you for the recommendation and review!

anonrig replied:
Somehow the following change causes "free(): invalid pointer"

```
void URLPattern::Test(const FunctionCallbackInfo<Value>& args) {
  URLPattern* url_pattern;
  ASSIGN_OR_RETURN_UNWRAP(&url_pattern, args.This());
  auto env = Environment::GetCurrent(args);

  ada::url_pattern_input input;
  std::optional<Utf8Value> input_base{};
  std::optional<Utf8Value> base_url{};
  if (args.Length() == 0) {
    input = ada::url_pattern_init{};
  } else if (args[0]->IsString()) {
    input_base = Utf8Value(env->isolate(), args[0].As<String>());
    input = input_base->ToStringView();
  } else if (args[0]->IsObject()) {
    input = URLPatternInit::FromJsObject(env, args[0].As<Object>());
  } else {
    THROW_ERR_INVALID_ARG_TYPE(
        env, "URLPattern input needs to be a string or an object");
    return;
  }

  if (args.Length() > 1) {
    if (!args[1]->IsString()) {
      THROW_ERR_INVALID_ARG_TYPE(env, "baseURL must be a string");
      return;
    }
    base_url = Utf8Value(env->isolate(), args[1].As<String>());
  }

  std::optional<std::string_view> base_url_opt =
      base_url ? std::optional(base_url->ToStringView()) : std::nullopt;
  args.GetReturnValue().Set(url_pattern->Test(env, input, base_url_opt));
}
```

addaleax replied:
Yeah, we should probably add `= delete` for the copy/move operators/constructors to `MaybeStackValue` to prevent people from running into this.

Using `input_base.emplace()` instead of `input_base = Utf8Value(...)` should work just fine.

addaleax commented on src/node_url_pattern.cc:
```diff
@@ -0,0 +1,791 @@
+    if (!ToV8Value(context, *init.base_url).ToLocal(&base_url)) {
+      return {};
+    }
+    USE(result->Set(context, env->base_url_string(), base_url));
```
I don't understand why we do proper error checking for `ToV8Value()` but skip it for `Set()` 🙂

jasnell replied:
Yep, these should be handled correctly. `USE()` is effectively just ignoring the possible error.

anonrig replied:
You're both correct. I hate the endless if statements that V8 is forcing me... Is there a better pattern to avoid this much if statement, or 1 if statement with a lot bool expressions?

addaleax replied:
> or 1 if statement with a lot bool expressions?

Yeah, I don't like it either but this is often the most convenient way to do this

addaleax commented on src/node_url_pattern.h:
```diff
@@ -0,0 +1,112 @@
+  bool Test(
+      Environment* env,
+      const ada::url_pattern_input& input,
+      std::optional<std::string_view>& baseURL);  // NOLINT (runtime/references)
```
Any reason to prefer NOLINT over the recommendation here?

(Although I feel like in this case this maybe should be a const reference anyway?)

anonrig replied:
If I make it a const parameter, then it doesn't compile due to this line:

```cpp
if (auto result = url_pattern_.exec(input, baseURL ? &*baseURL : nullptr)) {
```

exec() method in ada expects `std::string_view* base_url` where ada::url has a similar API surface as well, and making it const would differentiate the signature between url_pattern and url in Ada.

I'm not sure what is the best way, other than adding NOLINT and keeping this non-const.

jasnell replied:
blech... but yeah, that's better even if it looks ugly ;-)

legendecas commented on lib/internal/bootstrap/web/exposed-window-or-worker.js:
```diff
@@ -19,13 +19,18 @@ const {
 
+const { installObjectURLMethods, URLPattern } = require('internal/url');
+// https://urlpattern.spec.whatwg.org/#urlpattern-class
+exposeInterface(globalThis, 'URLPattern', URLPattern);
```
I noticed that the label https://github.com/nodejs/node/labels/semver-major was removed at https://github.com/nodejs/node/pull/56452#event-15892932080. Is it intentional?

anonrig replied:
I'll open another pull-request for making this global, rather than making this a semver-major, so we can backport it easily and land it on current release lines. This is on my todo list. I'll resolve this and remove the semver-major (once again) once I remove the global

anonrig replied:
Removed the global, will open another PR once this lands.

legendecas commented on src/node_url_pattern.cc:
```diff
@@ -0,0 +1,791 @@
+bool URLPatternRegexProvider::regex_match(std::string_view input,
+                                          const regex_type& pattern) {
+  auto isolate = Isolate::GetCurrent();
+  DCHECK_NOT_NULL(isolate);
```
nit: See https://github.com/nodejs/node/pull/56452/files#r1933001727, this DCHECK is not necessary.

anonrig replied:
Thanks. Removing.

jasnell commented:
PR has to be semver-major since it adds a new global

anonrig commented:
> PR has to be semver-major since it adds a new global

Removing semver-major. I've removed the global assignment to make this backportable to older release lines. I'll open another PR with semver-major label once this lands.

anonrig commented on lib/internal/bootstrap/web/exposed-window-or-worker.js:
```diff
@@ -19,13 +19,18 @@ const {
 
+const { installObjectURLMethods, URLPattern } = require('internal/url');
+// https://urlpattern.spec.whatwg.org/#urlpattern-class
+exposeInterface(globalThis, 'URLPattern', URLPattern);
```
Forgot to remove this.

lemire commented:
@anonrig At a glance, the Windows issues have to do with predefined macros clashing with variable names...

See https://github.com/ada-url/ada/pull/854 which should help.

anonrig commented:
> @anonrig At a glance, the Windows issues have to do with predefined macros clashing with variable names...
> 
> See [ada-url/ada#854](https://github.com/ada-url/ada/pull/854) which should help.

Thanks lemire. I rebased and updated Ada. Re-running tests now...

anonrig commented:
@nodejs/cpp-reviewers @nodejs/TSC appreciate it if you could review

benjamingr commented on doc/api/url.md:
```diff
@@ -714,6 +714,129 @@ Parses a string as a URL. If `base` is provided, it will be used as the base
+
+The constructor can throw a `TypeError` to indicate parsing failure.
+
+#### `new URLPattern(objg[, baseURL][, options])`
```
```suggestion
#### `new URLPattern(obj[, baseURL][, options])`
```

anonrig replied:
You have the eyes of an eagle.

ljharb commented on doc/api/url.md:
```diff
@@ -714,6 +714,129 @@ Parses a string as a URL. If `base` is provided, it will be used as the base
+
+Input can be a string or an object providing the individual URL parts. The
+object members can be any of `protocol`, `username`, `password`, `hostname`,
+`port`, `pathname`, `search`, `hash` or `baseURL`.
```
```suggestion
`port`, `pathname`, `search`, `hash`, or `baseURL`.
```

ljharb commented on doc/api/url.md:
```diff
@@ -714,6 +714,129 @@ Parses a string as a URL. If `base` is provided, it will be used as the base
+
+Input can be a string or an object providing the individual URL parts. The
+object members can be any of `protocol`, `username`, `password`, `hostname`,
+`port`, `pathname`, `search`, `hash` or `baseURL`.
```
```suggestion
`port`, `pathname`, `search`, `hash`, or `baseURL`.
```

ljharb commented on typings/internalBinding/url_pattern.d.ts:
```diff
@@ -0,0 +1,20 @@
+  search: string
+  hash: string
+
+  constructor(input: Record<string, string> | string, options?: { ignoreCase: boolean });
+  constructor(input: Record<string, string> | string, baseUrl?: string, options?: { ignoreCase: boolean });
```
```suggestion
  constructor(input: Record<string, string> | string, options?: { ignoreCase?: boolean });
  constructor(input: Record<string, string> | string, baseUrl?: string, options?: { ignoreCase?: boolean });
```
since an empty options object is still permitted

ljharb reviewed (approved):
I can't review the C++ but the parts I can look fine

anonrig commented:
@nodejs/tsc @nodejs/platform-smartos ... I can not justify the reasoning for waiting for smartOS build to succeed for more than 5.5 hours (https://ci.nodejs.org/job/node-test-commit-smartos/58956/nodes=smartos22-x64/) but here we are... The latest discussion didn't involve cases like this, we only had a discussion about flaky tests. 

**Please ask yourselves, if you were a new contributor, would you still continue contributing to Node.js after waiting for 5 hours for CI to succeed?**

bahamat commented:
@anonrig As you know, this is currently being discussed in nodejs/build#4011 (which you opened), and there are proposals there to correct the issue. I'd like to keep the main discussion in that issue so that the conversation is all in one place.

Thanks!

anonrig commented:
> @anonrig As you know, this is currently being discussed in [nodejs/build#4011](https://github.com/nodejs/build/issues/4011) (which you opened), and there are proposals there to correct the issue. I'd like to keep the main discussion in that issue so that the conversation is all in one place.
> 
> Thanks!

@bahamat That discussion doesn't solve the frustration of this PR right now. Any suggestions on how to land this pull-request now?

anonrig commented:
I stopped CI after 7 hours - https://ci.nodejs.org/job/node-test-pull-request/64872/

bahamat commented:
> @bahamat That discussion doesn't solve the frustration of this PR right now. Any suggestions on how to land this pull-request now?

I understand the frustration, but I don't have any suggestions. As per the TSC meeting, we have 48h (business days) to begin triage, which we have complied with here. This guideline was established for all platforms, not just SmartOS. We now have a remediation plan in place, which has been partially implemented, well ahead of the time guidelines.

You can follow the progress/discussion for remediation in nodejs/build#4011.

ShogunPanda reviewed (commented):
Not really qualified to review C++.
The rest LGTM. Nice job buddy!

Nimesh199398 commented:
> Review requested:
> 
> * [ ]  @nodejs/gyp
> * [ ]  @nodejs/security-wg
> * [ ]  @nodejs/startup
> * [ ]  @nodejs/url
> * [ ]  @nodejs/web-standards
